### ISSUE TITLE
Possible bug in DMC of the complex version on GPUs

### ISSUE THREAD

**Issue by maxamsler** [QMCPACK/qmcpack#219]
I believe I found a bug in QMCPack when using the complex version compiled for GPUs. The energies obtained with a DMC simulation in a periodic system with twists do not agree with the values obtained with the CPU version. This issue arises both with the double and mixed precision version of the GPU code.  I have attached the nexus file for a test case of a Si64 supercell. The twist averaged DMC energy for the CPU version results in:

qmca  -a -q ev -e 50  silicon/dmccpu_64_444/*001.scalar.dat
                            LocalEnergy               Variance           ratio 
avg  series 1  -252.106622 +/- 0.000573   1.820444 +/- 0.000643   0.0072 

This energy value is significantly lower than the value obtained from the GPU version:
qmca  -a -q ev -e 50  silicon/dmcgpu_64_444/*001.scalar.dat
                            LocalEnergy               Variance           ratio 
avg  series 1  -252.084855 +/- 0.000572   1.792653 +/- 0.000634   0.0071 

The bug does NOT seem to affect VMC calculations, since both CPU and GPU versions give results within the statistical error from each other.

The calculations were performed on Daint at the Swiss National Supercomputing center CSCS, a Cray XC50 with a NVIDIA Tesla P100 with 16GB per node. The following modules were loaded to compile QMCPack:
module switch  PrgEnv-cray/6.0.3 PrgEnv-intel
module switch intel/17.0.1.132 intel/16.0.3.210
module load cray-hdf5
module load daint-gpu
module load cudatoolkit

I configured the make as follows:
cmake -DCMAKE_C_COMPILER=cc -DCMAKE_CXX_COMPILER=CC -DQE_BIN=/project/s700/QMCPack/CPU/qmcpack/external_codes/quantum_espresso/espresso-5.3.0/bin -DQMC_COMPLEX=1  -DLibxml2_INCLUDE_DIRS=/project/s700/local/include/libxml2/ -DLIBXML2_LIBRARIES=/project/s700/local/ -DBOOST_INCLUDEDIR=/project/s700/boost_1_61_0/ -DQMC_INCLUDE="-I/project/s700/local/include/libxml2" -DQMC_EXTRA_LIBS="/project/s700/local/lib/libxml2.a /project/s700/local/lib/liblzma.a" -DQMC_CUDA=1 -DQMC_MIXED_PRECISION=0

[Si64.zip](https://github.com/QMCPACK/qmcpack/files/1015803/Si64.zip)

**Comment by prckent** [comment]
We are currently trying to reproduce this with smaller runs. As a first step #224 adds DMC tests for LiH at several twists/k-points in both locality and t-moves. We will run these nightly from now on, enabling comparisons between the CPU and GPU implementations, observing of any trends in reliability etc.

**Comment by prckent** [comment]
The LiH tests pass on GPU (Kepler w. Xeon host) with locality and t-moves and the standard mixed precision settings, indicating that this is a more subtle problem than "just" the DMC driver being broken.

**Comment by maxamsler** [comment]
I also observed the same discrepance in simple cubic phosphorus with 8 atoms, which is a smaller test case. If you want I can provide the reference data and input files.

**Comment by prckent** [comment]
Yes, that would help.

**Comment by maxamsler** [comment]
I have attached the nexus workflow together with the ".dat" files from the qmc calculations. The plot shows how the vmc/dmc energies as a function of excluded data for equilibration, for the CPU and GPU version of the code. The data for the plot and the gnuplot script are also attached in a separate tar file.
[plot.tar.gz](https://github.com/QMCPACK/qmcpack/files/1027297/plot.tar.gz)
[P_SC.tar.gz](https://github.com/QMCPACK/qmcpack/files/1027298/P_SC.tar.gz)

![plot](https://cloud.githubusercontent.com/assets/9530058/26426380/51b0d710-40a6-11e7-9ff1-e166716e38de.png)

**Comment by atillack** [comment]
Hello Max,

I've been going over your data and the one commonality that also leads to an offset in energy when I use our smaller LiH DMC tests is the usage of MPC in the Hamiltonian. I wonder if you could retest one set (CPU and GPU) with it off in your configuration files and if this fixes it.

Thanks,
Andreas

**Comment by ye-luo** [comment]
@atillack the energy value (local energy) reported by qmca doesn't include MPC correction. MPC correction is only an auxiliary Hamiltonian and doesn't affect sampling.

**Comment by atillack** [comment]
@ye-luo You're right. I reran the LiH tests with fixed random numbers and what I saw earlier (difference with MPC on, none with it off) went away. Wether MPC is used or not did not make a difference.

At least for the LiH DMC test, CPU and GPU complex runs agree within their error bars.

**Comment by maxamsler** [comment]
Ok. Has anyone tried to run the P-SC structure? Should run very quickly

**Comment by atillack** [comment]
@maxamsler I would like to. Could you please post a wave function and configuration file to save me the time compiling Quantum Espresso (on the P100 machine I am at it is being hard to compile right now).

**Comment by atillack** [comment]
@maxamsler The Nexus workflow you provided in your P_SC tarball does not work for me and fails with the following error:

`  Crystal error:
    the variable constants must be provided
  exiting.`

Could you please provide the wave function and configuration file?

**Comment by jtkrogel** [comment]
The problem with the workflow is a missing POSCAR file for the atomic structure.  Perhaps Max (@maxamsler) can add it here.

**Comment by prckent** [comment]
@jtkrogel Would be good for NEXUS to catch this.
@maxamsler We need the structure.

**Comment by maxamsler** [comment]
Sorry, I forgot to attach the POSCAR. 

[POSCAR.tar.gz](https://github.com/QMCPACK/qmcpack/files/1053845/POSCAR.tar.gz)
Here it is!

**Comment by atillack** [comment]
@maxamsler Thank you.

**Comment by jtkrogel** [comment]
@maxamsler I've tried to reproduce the initial parts of your workflow (orbital generation, jastrow optimization).  

I get the following for the scf total energy.  Can you confirm this is what you have?
psi1>grep '!  ' scf.out 
!    total energy              =    -104.92442028 Ry

For optimization, I get the following:
psi1>qmca -q ev *scalar*
 
                            LocalEnergy               Variance           ratio 
opt  series 0  -52.974221 +/- 0.011113   0.795427 +/- 0.017385   0.0150 
opt  series 1  -53.058415 +/- 0.009073   0.623527 +/- 0.013509   0.0118 
opt  series 2  -53.081506 +/- 0.012316   0.601191 +/- 0.015149   0.0113 
opt  series 3  -53.066947 +/- 0.008743   0.660168 +/- 0.010379   0.0124 
opt  series 4  -53.077711 +/- 0.010984   0.633892 +/- 0.011716   0.0119 
opt  series 5  -53.075013 +/- 0.007704   0.637929 +/- 0.009462   0.0120 
opt  series 6  -53.068671 +/- 0.005155   0.636694 +/- 0.009226   0.0120 
opt  series 7  -53.085528 +/- 0.006919   0.623939 +/- 0.008297   0.0118 

The supercell twist used in the optimization was:
Using supercell twist 0:  [  -0.12500  -0.12500  -0.12500]

We can proceed with the Jastrow I produced, but it would be better to have yours.  Can you post a single input file, e.g. dmc-8at_444/dmc.g000.twistnum_0.in.xml?

**Comment by maxamsler** [comment]
@jtkrogel Thank you. I get very similar results:
!    total energy              =    -104.92442028 Ry

And:
opt  series 0  -52.975754 +/- 0.013494   0.821427 +/- 0.015755   0.0155 
opt  series 1  -53.075402 +/- 0.004519   0.630866 +/- 0.007337   0.0119 
opt  series 2  -53.074534 +/- 0.005798   0.636403 +/- 0.005169   0.0120 
opt  series 3  -53.076695 +/- 0.005168   0.631304 +/- 0.005981   0.0119 
opt  series 4  -53.070720 +/- 0.004128   0.629416 +/- 0.004843   0.0119 
opt  series 5  -53.072013 +/- 0.003839   0.641991 +/- 0.004586   0.0121 
opt  series 6  -53.070086 +/- 0.004683   0.632449 +/- 0.003970   0.0119 
opt  series 7  -53.071325 +/- 0.003409   0.637983 +/- 0.004869   0.0120 

The dmc-8at_444/dmc.g000.twistnum_0.in.xml file is attached here: 
[dmc.g000.twistnum_0.in.xml.zip](https://github.com/QMCPACK/qmcpack/files/1065358/dmc.g000.twistnum_0.in.xml.zip)

I just realized that the workflow to produce the initial plot with the large difference between GPU and CPU was slightly different from the Nexus script I posted earlier. With the script posted above I get a smaller discrepancy between CPU and GPU, but still significant, as shown in the following plot:
![plot](https://user-images.githubusercontent.com/9530058/26996486-85083f4e-4d40-11e7-9df0-c99d346e1017.png)


However, in the workflow where the discrepancy was much larger, I used a small kgrid of 2x2x2 for the Jastrow optimization, which I then fed into the DMC calculations with a 3x3x3 and a 4x4x4 grid. Since the Jastrows/orbitals for both the GPU and CPU runs were identical I did not expect that it mattered, or am I wrong? It seems that the discrepancy between GPU and CPU depends on the Jastrow. Here is the workflow that optimizes the Jastrow on a 2x2x2 and runs QMC with a 3x3x3/4x4x4 grid, together with the Jastrow that I used for the DMC runs. 
[P_SC_loop.py.zip](https://github.com/QMCPACK/qmcpack/files/1065263/P_SC_loop.py.zip)
[dmc.g000.twistnum_0.in.xml.zip](https://github.com/QMCPACK/qmcpack/files/1065335/dmc.g000.twistnum_0.in.xml.zip)

**Comment by prckent** [comment]
We have found what looks to be a definite bug in the DMC GPU code and its handling of certain twists. A DMC run with an x-point twist reliably gives a different result to the CPU code. https://cdash.qmcpack.org/CDash/testDetails.php?test=506898&build=7683 The equivalent VMC run is OK and the CPU and GPU runs are consistent. This may not be the same problem as the bug discussed here, but there could well be some relation.

**Comment by jtkrogel** [comment]
Does this bug manifest in real and complex gpu, or complex only?

**Comment by prckent** [comment]
This "new" bug affects only the real build of the GPU code. The newly written complex code is correct https://cdash.qmcpack.org/CDash/testDetails.php?test=507363&build=7687

**Comment by atillack** [comment]
@prckent @jtkrogel We should probably open a new bug for this. Attached find a plot of the differences for the LiH-x-short runs:

![results-lih-short-x](https://user-images.githubusercontent.com/24940649/27187962-a2d047c8-51ba-11e7-8cd1-f6a06139650e.png)

The real wave function GPU runs are consistently off while the CPU and GPU (complex) runs are consistent with each other and the reference value.

**Comment by atillack** [comment]
@maxamsler Here is a (slightly hard to read) plot of your phosphorus example data at each twist number comparing CPU and GPU runs overlaid with reruns (for good measure four times the number of blocks, 400, and with a four times smaller time step of 0.01) of the same wave functions at select twists (every four) to spot check. For your runs I averaged over the last 30 blocks using qmca and for the four times longer spot checks over the last 120.

![results_psc](https://user-images.githubusercontent.com/24940649/27189792-6ec826d4-51c0-11e7-8384-e883d888c33b.png)

Long story short, while I do see the occasional data point being off by more than 1 sigma (2 sigma is the most I've seen) there is nothing I would consider dramatically off.

**Comment by jtkrogel** [comment]
@atillack Can you also post the twist averaged final numbers (qmca -a ...) in text format for CPU/GPU, Max/Rerun?

**Comment by atillack** [comment]
@jtkrogel Great suggestion, I think we're getting somewhere... Let me explain:

Here are the values for the last 30 blocks for Max' data:
Overall CPU:
avg  series 1  -52.837026 +/- 0.000367   0.622537 +/- 0.000317   0.0118 
Overall GPU:
avg  series 1  -52.835842 +/- 0.000392   0.620309 +/- 0.000297   0.0117

And for the rerun for the last 120:
Overall CPU:
avg  series 1  -53.016886 +/- 0.000634   0.622187 +/- 0.000545   0.0117 
Overall GPU:
avg  series 1  -53.017832 +/- 0.000646   0.622489 +/- 0.000557   0.0117 

Keep in mind that for the rerun I am only averaging over every fourth twist.

This is interesting as the calculated errors by qmca are about an order of magnitude smaller compared to the errors reported for the individual twists, for Max' data:

CPU (twist #0): -53.196303 +/- 0.002710 0.619067 +/- 0.002518 0.0116
GPU (twist #0): -53.197061 +/- 0.001960 0.613733 +/- 0.002728 0.0115

For the reruns:
CPU (twist #0): -53.200358 +/- 0.003109 0.619121 +/- 0.002440 0.0116
GPU (twist #0): -53.197160 +/- 0.002041 0.620312 +/- 0.003893 0.0117

How does that explain anything you might ask... What qmca is doing is it calculates the error by assuming all data points come from the same distribution, in other words, the variance drops by 1/n (n being the number of samples).

When averaging over different twists with a complex wave function this assumption does not hold anymore (see the plot I posted earlier, with very discrete energy levels) and hence the error obtained with qmca is overly optimistic. Off the cuff, a simple correction (as such likely too naive) for the phosphorus case with four energy levels could be to multiply qmca's error with sqrt(4)...

**Comment by jtkrogel** [comment]
So there are two potential issues:  

1) The error bars may be estimated poorly due to the low number of blocks involved.  The errorbar has an errorbar and I would recommend using upwards of 200 blocks to calculate it correctly.  

2) The use of gaussian statistics in general. qmca, and all other qmc statistical processing tools I am aware of, make the partially poor assumption of gaussian statistics.  This is true for single point or twist averaged calculations, but as you point out in the twist averaged case, the averaging itself makes an additional reliance on gaussian statistics (gaussian variance averaging formula), which may further expose the weakness of this assumption.   

One way to correct for the additional assumption embedded in 2) for the case of twist averaging is to directly twist average the data files and then perform the statistics (difficulties arise when the files differ in length).  I have an older tool that does this.  @atillack If you post all the relevant scalar.dat files here, I will investigate this difference (though the insufficient blocks problem (1) may remain).

**Comment by prckent** [comment]
@atillack Very interesting that there is no big difference at the individual twist level. Are you certain? Looking at the plot, I can't be certain that there is not a missing point.

**Comment by atillack** [comment]
@jtkrogel Here are all the *.scalar.dat files from Max and my rerun:

[scalar_dat_files.zip](https://github.com/QMCPACK/qmcpack/files/1078530/scalar_dat_files.zip)

@prckent You can look into the two text files I posted containing the data the plot is made from. The largest deviation between CPU and GPU I see in Max' data is for twist 34 with a 3 sigma deviation. In the rerun it is for twist 32 with a 2 sigma deviation.

**Comment by jtkrogel** [comment]
OK, here's my summary.

40 blocks (Max's data, post equilibration) is too few to calculate a reliable error bar for 2 reasons: 1) large variation in error bar due to too few blocks, 2) more importantly, too few blocks leads to an underestimated autocorrelation time which then leads to an underestimated errorbar.  Overall, there is no (statistically) significant evidence for a difference between cpu and gpu runs.

Max's results (twist averaged, last 40 blocks): 
maui>qmca -a -e 60 -q e dmc.*s001*scalar*  --sac
avg  series 1  LocalEnergy           =  -52.836690 +/- 0.000445    2.5 
maui>qmca -a -e 60 -q e dmcgpu.*s001*scalar*  --sac
avg  series 1  LocalEnergy           =  -52.835522 +/- 0.000510    3.1 

difference:   -0.0012    +/- 0.0007  (1.5 "sigma", underestimated error bar)
autocorrelation time: ~3 blocks

Andrea's longer runs (twist averaged, last 200 blocks)
maui>qmca -a -e 200 -q e *cpu_long*s001*scalar*  --sac
avg  series 1  LocalEnergy           =  -53.016785 +/- 0.000804    7.5 
maui>qmca -a -e 200 -q e *gpu_long*s001*scalar*  --sac
avg  series 1  LocalEnergy           =  -53.018113 +/- 0.000841    8.5 

difference:  0.0013   +/- 0.0011 (1 sigma, or essentially zero)
autocorrelation time: ~8 blocks

In either set of runs, the VMC data is too short (20 blocks) to get a meaningful errorbar.


On point 2) of my previous post, twist averaged error bars seem to be estimated accurately under the current assumptions:

Twist averaged error bar for Andreas' cpu runs: 0.000804
Amplify this by sqrt(17) (17 files in the average): 0.003315
Average over individual error bars for each twist: 0.003207

In other words, the errorbar expected for a single twist (based on the TA estimate), matches the observed behavior of a single twist.

**Comment by jtkrogel** [comment]
Forgot to mention, averaging full file data then doing stats gives an error bar of 0.000789, while estimating the errorbar via the variance averaging formula gives 0.000804.

**Comment by prckent** [comment]
@maxamsler For now, our conclusion is that this is not a bug but a consequence of too few blocks in your runs. This gives unreliable error bars (and error of error bars). We suggest you try, e.g. 200 blocks leaving the step and walker count unchanged and the problem should go away. Please let us know. If the problem remains, then we will revisit it. The last uncontrolled variable is the CSCS machine.

**Comment by prckent** [comment]
Will reopen if necessary.

**Comment by maxamsler** [comment]
Thanks a lot. I am running now with 300 blocks to see if the error bars decrease.

Best
Max

**Comment by maxamsler** [comment]
Hi Paul,
I just redid the calculations using 800 blocks, and I still get some discrepancies. However, I did not change how the errors were computed. Should I multiply the errorbars with sqrt(ntwist)? Otherwise, it seems to be an issue with my computer cluster.


Best
Max

**Comment by jtkrogel** [comment]
@maxamsler I would like to take a closer look at your 800 block output files.  If you could post them here, that would be great.

**Comment by maxamsler** [comment]
Here are the files, incl. the plot. 

![plot eps](https://user-images.githubusercontent.com/9530058/27297218-064abb0a-54f2-11e7-8621-4de75c6fa159.png)


[800blocks_dat_gpu.tar.gz](https://github.com/QMCPACK/qmcpack/files/1085763/800blocks_dat_gpu.tar.gz)
[800blocks_dat_cpu.tar.gz](https://github.com/QMCPACK/qmcpack/files/1085765/800blocks_dat_cpu.tar.gz)

**Comment by jtkrogel** [comment]
@maxamsler Earlier I missed the fact that your runs were performed with timestep 0.04 while Andreas' runs used 0.01.  This alone explains the difference between the autocorrelation times.

We will rerun at 0.04 to see if we reproduce the difference you see; it is possible that this reflects differing timestep behavior between the cpu and gpu implementations.  If so, moving to a smaller timestep (~0.01) might resolve your issue (implementations are correct so long as they agree in the zero timestep limit).

**Comment by maxamsler** [comment]
Here are the results for runs with time step of 0.01. This time, I also increased the number of blocks to 2000. Nevertheless, I get a significant discrepancy in the energies.
![plot eps](https://user-images.githubusercontent.com/9530058/27399658-850a268e-568b-11e7-947e-7f687faf95a7.png)

I had to split the data into smaller chunks:
[data_2000.tar.gz.partag.txt](https://github.com/QMCPACK/qmcpack/files/1092497/data_2000.tar.gz.partag.txt)
[data_2000.tar.gz.partaf.txt](https://github.com/QMCPACK/qmcpack/files/1092498/data_2000.tar.gz.partaf.txt)
[data_2000.tar.gz.partae.txt](https://github.com/QMCPACK/qmcpack/files/1092500/data_2000.tar.gz.partae.txt)
[data_2000.tar.gz.partad.txt](https://github.com/QMCPACK/qmcpack/files/1092501/data_2000.tar.gz.partad.txt)
[data_2000.tar.gz.partac.txt](https://github.com/QMCPACK/qmcpack/files/1092502/data_2000.tar.gz.partac.txt)
[data_2000.tar.gz.partab.txt](https://github.com/QMCPACK/qmcpack/files/1092499/data_2000.tar.gz.partab.txt)
[data_2000.tar.gz.partaa.txt](https://github.com/QMCPACK/qmcpack/files/1092503/data_2000.tar.gz.partaa.txt)

**Comment by prckent** [comment]
Changed title since the current hypothesis is that this problem is probably statistical. It could also be related to the CSCS machine (software, Pascal?). For a single run the twists are individually close enough when run on OLCF Kepler machines. Multiple runs to obtain robust statistics and twist by twist comparisons could shed more light.

**Comment by prckent** [comment]
I am going to close this issue since we believe this to be statistical problem intrinsic to the setup of the twists in this case and not a bug in QMCPACK. i.e. It is a hazard of statistical methods. Our comparisons between the CPU and GPU runs for individual twists agree. We can reopen this issue if needed. It does point to the need for better statistical analysis procedures.
